src/dbnode/storage/index/block.go

```diff
@@ -102,6 +106,61 @@ func (s blockState) String() string {
 	return "unknown"
 }
 
+// LimitBlocksQueried enforces a max blocks query limit within a recency window.
+func LimitBlocksQueried(max int, within time.Duration) chan struct{} {
+	recentlyQueried = &queryRecencyWindow{
+		length:         within,
+		maxBlocks:      max,
+		recentBlocks:   atomic.NewInt64(0),
+		previousBlocks: 0,
+		stopCh:         make(chan struct{}),
+	}
+	recentlyQueried.Start()
+
+	// Give caller handle to stop the background flushing.
+	return recentlyQueried.stopCh
+}
+
+// For tracking query stats in past X duration such as blocks queried.
+type queryRecencyWindow struct {
+	length    time.Duration
+	maxBlocks int
+
+	recentBlocks   *atomic.Int64
+	previousBlocks int64
+	stopCh         chan struct{}
+}
+
+func (w queryRecencyWindow) Start() {
+	ticker := time.NewTicker(w.length)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			// Clear older active blocks every X duration.
+			w.recentBlocks.Sub(w.previousBlocks)
```

Comment:
Best to store(0)

Reply:
Ah yeah you're right. No need to add/subtract here; can just reset to zero.